Improve MISSING shards snapshot message

DaveCTurner opened the pull request:
Today if you try and take a non-partial snapshot of some indices whose
primaries are not all assigned you get a failure message saying

    Indices don't have primary shards

This is fairly confusing and suggests that the user might need to try
and add primary shards like they can add replicas. In fact the indices
_do_ have primary shards, they're just unassigned. This commit clarifies
the message, and adds a link to the relevant troubleshooting docs.

DaveCTurner committed:
Improve MISSING shards snapshot message

Today if you try and take a non-partial snapshot of some indices whose
primaries are not all assigned you get a failure message saying

    Indices don't have primary shards

This is fairly confusing and suggests that the user might need to try
and add primary shards like they can add replicas. In fact the indices
_do_ have primary shards, they're just unassigned. This commit clarifies
the message, and adds a link to the relevant troubleshooting docs.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner committed:
Mention partial

idegtiarenko commented:
https://github.com/elastic/elasticsearch/blob/414eda7b801372a3d1a98c9d25dd2bad4572abad/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CreateSnapshotStep.java#L129

This part looks a little weird to me. Is it worth updating this message as well?
May be `failed to create snapshot, % out of % total shard(s) failed`?

ywangd commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -4024,7 +4025,11 @@ private SnapshotsInProgress createSnapshot(
+                        the following indices have unassigned primary shards \
+                        and cannot be included in a snapshot with [partial] set to [false]: %s; \
+                        for help with troubleshooting unassigned shards see %s
+                        """, missing, ReferenceDocs.UNASSIGNED_SHARDS));
```
Nit: might be helpful to sort `missing` for error message.

ywangd commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -4024,7 +4025,11 @@ private SnapshotsInProgress createSnapshot(
-                    throw new SnapshotException(snapshot, "Indices don't have primary shards " + missing);
+                    throw new SnapshotException(snapshot, Strings.format("""
+                        the following indices have unassigned primary shards \
+                        and cannot be included in a snapshot with [partial] set to [false]: %s; \
```
`partial` defaults to `false` so that users get this behaviour without explicitly "set" it to `false`. I wonder whether we should provide some extra hint here?

ywangd reviewed (approved):
LGTM

DaveCTurner commented:
> Is it worth updating this message as well?

Mmm possibly although it seems kinda unrelated to this PR.

I'm a little surprised you can even hit this message given that the snapshot in question has `partial: false`: the comment says that an aborted snapshot might still succeed with some failed shards, and in practice we do see this message in production, but I would have expected a full snapshot to just fail outright if it doesn't completely succeed. TIL.

DaveCTurner committed:
Review feedback

- Reverse sense of `[partial]`: `false` is the default so may not really be set
- Sort indices, and add test for this

DaveCTurner committed:
Merge branch 'main' into 2024/02/14/snapshot-unassigned-shards-message

DaveCTurner committed:
Reformat

DaveCTurner committed:
Merge branch 'main' into 2024/02/14/snapshot-unassigned-shards-message
